flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/ProcTimeInnerJoin.scala

```diff
@@ -0,0 +1,323 @@
+    if (winSize != 0) {
+      // register a timer to expire the element
+      if (timerState.value == 0 && winSize != -1) {
+        ctx.timerService.registerProcessingTimeTimer(curProcessTime + winSize + 1)
+        timerState.update(curProcessTime + winSize + 1)
+      }
+
+      var rowList = rowMapState.get(curProcessTime)
+      if (rowList == null) {
+        rowList = new util.ArrayList[Row]()
+      }
+      rowList.add(value)
+      rowMapState.put(curProcessTime, rowList)
+
+    }
+
+    // loop the rightstream elments
+    val oppositeKeyIter = oppoRowMapState.keys().iterator()
+    while (oppositeKeyIter.hasNext) {
+      val eleTime = oppositeKeyIter.next()
+      if (eleTime < oppoExpiredTime) {
+        listToRemove.add(eleTime)
+      } else {
+        val oppoRowList = oppoRowMapState.get(eleTime)
+        var i = 0
+        while (i < oppoRowList.size) {
+          if (isPositive) {
+            compositeOutput(value, oppoRowList.get(i))
+          } else {
+            compositeOutput(oppoRowList.get(i), value)
+          }
+
+          if (filterFunc.filter(outputC.row)) {
+            out.collect(outputC)
+          }
+          i += 1
+        }
+      }
+    }
+
+    // expire records out-of-time
+    var i = listToRemove.size() - 1
+    while (i >= 0) {
+      oppoRowMapState.remove(listToRemove.get(i))
+      listToRemove.remove(i)
+      i -= 1
+    }
+  }
+
+  /**
+    * set output fields according left and right stream row
+    */
+  private def compositeOutput(
+     leftRow: Row,
+     rightRow: Row): Unit = {
+
+    var i = 0
+    while (i < element1Type.getArity) {
+      outputC.row.setField(i, leftRow.getField(i))
+      i += 1
+    }
+
+    i = 0
+    while (i < element2Type.getArity) {
+      outputC.row.setField(i + element1Type.getArity, rightRow.getField(i))
+      i += 1
+    }
+
+  }
+
+  /**
+    * expire records which before curTime - windowSize,
+    * and register a timer if still exist records.
+    * Ensure that one key only has one timer, so register another
+    * timer until last timer trigger.
+    */
+  private def expireOutTimeRow(
+    curTime: Long,
+    winSize: Long,
+    rowMapState: MapState[Long, JList[Row]],
+    timerState: ValueState[Long],
+    ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext): Unit = {
+
+    val expiredTime = curTime - winSize
+    val expiredList = new util.ArrayList[Long]
+    val keyIter = rowMapState.keys().iterator()
+    var nextTimer: Long = 0
+    // loop the timestamps to find out expired records, when meet one record
+    // after the expried timestamp, break the loop. If the keys is ordered, thus
+    // can reduce loop num, if the keys is unordered, also can expire at least one
+    // element every time the timer trigger
+    while (keyIter.hasNext && nextTimer == 0) {
+      val curTime = keyIter.next
+      if (curTime < expiredTime) {
+        expiredList.add(curTime)
+      } else {
+        nextTimer = curTime
```

Comment:
Yes, when it found an record after the expired time, i will not continue to loop, it will register timer for this record to delete itself until next timer trigger, thus can reduce useless loop operation.